File: agents/alphacapture.py
# program uses minimax algorithm with alpha-beta pruning to play Mancala
# AlphaCapture will determine the best move seqeuence as the "maximizing" player

# creates AlphaCapture class
# sets max player and depth for search
class AlphaCapture:
    def __init__(self, max_player=0, depth=12):
        self.MAX_PLAYER = max_player
        self.depth = depth

    # possible actions in a state
    def actions(self, state):
        board, player = state
        if player == 0:
            return [i for i in range(0, 6) if board[i] > 0]
        else:
            return [i for i in range(7, 13) if board[i] > 0]
    
    # result of applying an action to a state
    def result(self, state, action):
        board, player = state
        new_board = board[:]
        stones = new_board[action]
        new_board[action] = 0
        index = action
        while stones > 0:
            index = (index + 1) % 14
            if (player == 0 and index == 13) or (player == 1 and index == 6):
                continue
            new_board[index] += 1
            stones -= 1
        # Check for capture
        player_pits = range(0, 6) if player == 0 else range(7, 13)
        if index in player_pits and new_board[index] == 1:
            opposite = 12 - index
            if new_board[opposite] > 0:
                mancala = 6 if player == 0 else 13
                new_board[mancala] += new_board[opposite] + 1
                new_board[opposite] = 0
                new_board[index] = 0
        # Check for extra turn
        mancala_index = 6 if player == 0 else 13
        extra_turn = (index == mancala_index)
        new_player = player if extra_turn else 1 - player
        return (new_board, new_player)
    
    # terminal test for a state
    def terminal(self, state):
        board = state[0]
        return sum(board[0:6]) == 0 or sum(board[7:13]) == 0

    # utility of a terminal state
    def utility(self, state):
        board = state[0][:]
        if sum(board[0:6]) == 0:
            board[13] += sum(board[7:13])
            for i in range(7, 13): board[i] = 0
        elif sum(board[7:13]) == 0:
            board[6] += sum(board[0:6])
            for i in range(0, 6): board[i] = 0
        
        if self.MAX_PLAYER == 0:
            return board[6] - board[13]
        else:
            return board[13] - board[6]

    # heuristic evaluation for non-terminal states
    def heuristic(self, state):
        board = state[0]

        # determine MAX (AI) vs MIN (opponent)
        if self.MAX_PLAYER == 0:
            my_store, opp_store = board[6], board[13]
            my_pits = range(0, 6)
            opp_pits = range(7, 13)
            my_skip = 13   # skip opponent store
            opp_skip = 6
        else:
            my_store, opp_store = board[13], board[6]
            my_pits = range(7, 13)
            opp_pits = range(0, 6)
            my_skip = 6
            opp_skip = 13

        # store difference (most important term)
        score = 3 * (my_store - opp_store)

        # capture potential for MAX (AI)
        capture_bonus = 0
        for pit in my_pits:
            stones = board[pit]
            if stones == 0:
                continue
            landing = pit
            s = stones
            while s > 0:
                landing = (landing + 1) % 14
                if landing == my_skip:
                    continue
                s -= 1
            if landing in my_pits and board[landing] == 0:
                capture_bonus += board[12 - landing]

        # capture potential for MIN (opponent)
        capture_penalty = 0
        for pit in opp_pits:
            stones = board[pit]
            if stones == 0:
                continue
            landing = pit
            s = stones
            while s > 0:
                landing = (landing + 1) % 14
                if landing == opp_skip:
                    continue
                s -= 1
            if landing in opp_pits and board[landing] == 0:
                capture_penalty += board[12 - landing]

        return score + 2 * capture_bonus - 2 * capture_penalty

    # minimax algorithm with depth limit and alpha-beta pruning
    def minimax(self, state, depth, alpha, beta):
        if self.terminal(state):
            return self.utility(state)
        if depth == 0:
            return self.heuristic(state)

        player = state[1]

        if player == self.MAX_PLAYER:  # MAX node
            value = float('-inf')
            for action in self.actions(state):
                value = max(
                    value, self.minimax(self.result(state, action), depth - 1, alpha, beta))
                alpha = max(alpha, value)
                if alpha >= beta:
                    break
            return value

        else:  # MIN node
            value = float('inf')
            for action in self.actions(state):
                value = min(
                    value, self.minimax(self.result(state, action), depth - 1, alpha, beta))
                beta = min(beta, value)
                if beta <= alpha:
                    break
            return value

    # choose the best action using minimax
    def best_action(self, state):
        best_value = float('-inf')
        best_action = None

        for action in self.actions(state):
            value = self.minimax(self.result(state, action), self.depth - 1, float('-inf'), float('inf'))
            if value > best_value:
                best_value = value
                best_action = action

        return best_action
    
    # choose the best sequence of actions using minimax (if possible extra turns)
    def best_sequence(self, state):
    
        sequence = []
        current_state = state

        while True:
            _, player = current_state
            # stop if game is over or it's not MAX's turn
            if self.terminal(current_state) or player != self.MAX_PLAYER:
                break
            best_value = float('-inf')
            best_action = None
            for action in self.actions(current_state):
                value = self.minimax(self.result(current_state, action), self.depth - 1, float('-inf'), float('inf'))
                if value > best_value:
                    best_value = value
                    best_action = action

            # apply the best action
            sequence.append(best_action)
            next_state = self.result(current_state, best_action)

            # stop chaining if turn switches to opponent
            if next_state[1] != self.MAX_PLAYER:
                break

            current_state = next_state

        return sequence

# Example usage:
if __name__ == "__main__":
    initial_board = [4, 4, 4, 4, 4, 4, 0,
                     4, 4, 4, 4, 4, 4, 0]
    initial_state = (initial_board, 0)  # Player 0 starts

    bot = AlphaCapture(max_player=0, depth=8)
    best_seq = bot.best_sequence(initial_state)
    print("Best move sequence for Player 0:", best_seq)

File: web/src/logic/alphacapture.js
// Minimax with alpha-beta pruning for Mancala.
// Board layout: indices 0-5 = P0 pits, 6 = P0 store, 7-12 = P1 pits, 13 = P1 store.

export function actions(state) {
  const [board, player] = state;
  if (player === 0) return [0,1,2,3,4,5].filter(i => board[i] > 0);
  return [7,8,9,10,11,12].filter(i => board[i] > 0);
}

export function result(state, action) {
  const [board, player] = state;
  const b = board.slice();
  let stones = b[action];
  b[action] = 0;
  let index = action;
  while (stones > 0) {
    index = (index + 1) % 14;
    if (player === 0 && index === 13) continue;
    if (player === 1 && index === 6) continue;
    b[index]++;
    stones--;
  }
  // capture
  const myPits = player === 0 ? [0,1,2,3,4,5] : [7,8,9,10,11,12];
  if (myPits.includes(index) && b[index] === 1) {
    const opposite = 12 - index;
    if (b[opposite] > 0) {
      const store = player === 0 ? 6 : 13;
      b[store] += b[opposite] + 1;
      b[opposite] = 0;
      b[index] = 0;
    }
  }
  // extra turn
  const mancala = player === 0 ? 6 : 13;
  const nextPlayer = index === mancala ? player : 1 - player;
  return [b, nextPlayer];
}

export function terminal(state) {
  const board = state[0];
  return board.slice(0,6).every(x => x === 0) || board.slice(7,13).every(x => x === 0);
}

function utility(state, maxPlayer) {
  const b = state[0].slice();
  if (b.slice(0,6).every(x => x === 0)) {
    b[13] += b.slice(7,13).reduce((a,x) => a+x, 0);
    for (let i = 7; i < 13; i++) b[i] = 0;
  } else {
    b[6] += b.slice(0,6).reduce((a,x) => a+x, 0);
    for (let i = 0; i < 6; i++) b[i] = 0;
  }
  return maxPlayer === 0 ? b[6] - b[13] : b[13] - b[6];
}

function heuristic(state, maxPlayer) {
  const board = state[0];
  const myStore  = maxPlayer === 0 ? board[6]  : board[13];
  const oppStore = maxPlayer === 0 ? board[13] : board[6];
  const myPits  = maxPlayer === 0 ? [0,1,2,3,4,5]   : [7,8,9,10,11,12];
  const oppPits = maxPlayer === 0 ? [7,8,9,10,11,12] : [0,1,2,3,4,5];
  const mySkip  = maxPlayer === 0 ? 13 : 6;
  const oppSkip = maxPlayer === 0 ? 6  : 13;

  let score = 3 * (myStore - oppStore);

  let captureBonus = 0;
  for (const pit of myPits) {
    let stones = board[pit];
    if (!stones) continue;
    let landing = pit;
    let s = stones;
    while (s > 0) {
      landing = (landing + 1) % 14;
      if (landing === mySkip) continue;
      s--;
    }
    if (myPits.includes(landing) && board[landing] === 0) {
      captureBonus += board[12 - landing];
    }
  }

  let capturePenalty = 0;
  for (const pit of oppPits) {
    let stones = board[pit];
    if (!stones) continue;
    let landing = pit;
    let s = stones;
    while (s > 0) {
      landing = (landing + 1) % 14;
      if (landing === oppSkip) continue;
      s--;
    }
    if (oppPits.includes(landing) && board[landing] === 0) {
      capturePenalty += board[12 - landing];
    }
  }

  return score + 2 * captureBonus - 2 * capturePenalty;
}

function minimax(state, depth, alpha, beta, maxPlayer, stats) {
  stats.nodes++;
  if (terminal(state)) return utility(state, maxPlayer);
  if (depth === 0) return heuristic(state, maxPlayer);

  const player = state[1];
  const moves = actions(state);

  if (player === maxPlayer) {
    let value = -Infinity;
    for (const action of moves) {
      value = Math.max(value, minimax(result(state, action), depth - 1, alpha, beta, maxPlayer, stats));
      alpha = Math.max(alpha, value);
      if (alpha >= beta) break;
    }
    return value;
  } else {
    let value = Infinity;
    for (const action of moves) {
      value = Math.min(value, minimax(result(state, action), depth - 1, alpha, beta, maxPlayer, stats));
      beta = Math.min(beta, value);
      if (beta <= alpha) break;
    }
    return value;
  }
}

// Returns { action, stats: { nodes, ms, value } }
export function bestAction(state, maxPlayer = 0, depth = 8) {
  const stats = { nodes: 0 };
  const t0 = performance.now();
  let bestValue = -Infinity;
  let bestAct = null;
  for (const action of actions(state)) {
    const value = minimax(result(state, action), depth - 1, -Infinity, Infinity, maxPlayer, stats);
    if (value > bestValue) { bestValue = value; bestAct = action; }
  }
  stats.ms = Math.round(performance.now() - t0);
  stats.value = bestValue;
  return { action: bestAct, stats };
}

// Chains extra-turn moves. Returns { sequence, stats }
export function bestSequence(state, maxPlayer = 0, depth = 8) {
  const sequence = [];
  let current = state;
  const totalStats = { nodes: 0, ms: 0, value: 0 };
  while (true) {
    const [, player] = current;
    if (terminal(current) || player !== maxPlayer) break;
    const { action, stats } = bestAction(current, maxPlayer, depth);
    if (action === null) break;
    totalStats.nodes += stats.nodes;
    totalStats.ms += stats.ms;
    totalStats.value = stats.value;
    sequence.push(action);
    const next = result(current, action);
    if (next[1] !== maxPlayer) break;
    current = next;
  }
  return { sequence, stats: totalStats };
}
